### PR TITLE
[WIP] Use topic ID instead of topic hash

### PR DESCRIPTION
**Pull request by jamesray1**

### CODE REVIEW

**Comment by jamesray1** [comment]
Great, the tests passed. @mgoelzer and @tomaka, do you want to review and test this?

File: protocols/floodsub/src/lib.rs [hunk 1]
```diff
@@ -391,7 +392,7 @@ impl FloodSubController {
         let topics = topics.into_iter().collect::<Vec<_>>();
 
         debug!("Queueing publish message ; topics = {:?} ; data_len = {:?}",
-               topics.iter().map(|t| t.hash().clone().into_string()).collect::<Vec<_>>(),
+               topics.iter().map(|t| t.descriptor.get_name().to_string()).collect::<Vec<_>>(),
```
Comment (rphmeier) [inline protocols/floodsub/src/lib.rs]: is allocation necessary here?

File: protocols/floodsub/src/lib.rs [hunk 2]
```diff
@@ -635,7 +635,8 @@ fn handle_packet_received(
             let subscribed_topics = inner.subscribed_topics.read();
             topics
                 .iter()
-                .any(|t| subscribed_topics.iter().any(|topic| topic.hash() == t))
+                .any(|t| subscribed_topics.iter().any(|topic| 
+                    topic.descriptor.clone().get_name().to_string() == *t))
```
Comment (rphmeier) [inline protocols/floodsub/src/lib.rs]: does allocation need to be done here? can't we check equality without reallocating?

File: protocols/floodsub/src/topic.rs [hunk 3]
```diff
@@ -47,7 +47,7 @@ impl TopicHash {
 /// Built topic.
 #[derive(Debug, Clone)]
 pub struct Topic {
-    descriptor: rpc_proto::TopicDescriptor,
+    pub descriptor: rpc_proto::TopicDescriptor,
```
Comment (rphmeier) [inline protocols/floodsub/src/topic.rs]: Docs on public field

Comment (tomaka) [reply protocols/floodsub/src/topic.rs]: Also this shouldn't be public. There should be a getter like `name(&self) -> &str`.
If we stop using hashes altogether, then the whole content of the module may need some revisiting.

Comment (jamesray1) [reply protocols/floodsub/src/topic.rs]: OK, will implement a getter tomorrow in Sydney.

File: protocols/floodsub/src/lib.rs (hunk 2 above)
Comment (jamesray1) [reply protocols/floodsub/src/lib.rs]: I'll look into that; although I'm not sure how it might be done off the top of my head.

**Comment by mgoelzer** [comment]
@jamesray1 This will completely remove the topic hash - am I understanding that right?  I don't think there is a consensus in https://github.com/libp2p/rust-libp2p/issues/473 that that's the right solution here.  My concern is that rust-libp2p did the right thing, and we're going to revert it to accommodate some brokenness in go-libp2p and js-libp2p.

**Comment by jamesray1** [comment]
@mgoelzer the topic hash will still be available to use, although where it has been used I have changed it to use the topic ID instead. But I did also think that using the topic hash seemed reasonable, and I will wait for consensus in #473.

**Comment by jamesray1** [comment]
Closing as doing this as part of https://github.com/libp2p/rust-libp2p/pull/521.

**Comment by jamesray1** [comment]
Then again it will be easier to review if it is separate from gossipsub...

**Comment by jamesray1** [comment]
I'll just do this in #521.

**Pull request closed without merging**